File: src/eventdetection/alignment.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"Cycle alignment: define an absolute zero common to all cycles"
from   typing import (Union, Optional, # pylint: disable=unused-import
                      Sequence, Iterable, Iterator, cast)
from   enum   import Enum
import numpy  as     np
from   numpy.lib.index_tricks import as_strided

from   utils                    import initdefaults, kwargsdefaults, changefields
from   signalfilter.convolve    import KernelConvolution

class AlignmentMode(Enum):
    "Computation modes ExtremumAlignment."
    min = 'min'
    max = 'max'

class ExtremumAlignment:
    """
    Functor which an array of biases computed as the extremum of provided ranges.
    Biases are furthermore centered at zero around their median

    Attributes:

    * *mode*: the extremum to use
    * *binsize*: if > 2, the extremum is computed over the median of values binned
        by *binsize*.
    """
    binsize = 5
    mode    = AlignmentMode.min
    @initdefaults(frozenset(locals()))
    def __init__(self, **_):
        pass

    def __setstate__(self, kwa):
        self.__init__(**kwa)

    def __get(self, elem):
        if len(elem) <= 2:
            return np.NaN

        elif len(elem) <= self.binsize:
            return np.median(elem)

        bsize  = self.binsize
        binned = as_strided(elem,
                            shape   = (len(elem)-bsize+1, bsize),
                            strides = (elem.strides[0],)*2)
        return np.median(binned, axis = 0)

    def one(self, data) -> np.ndarray:
        "call on one table"
        fcn = getattr(np, self.mode.value)
        return -fcn(self.__get(data) if self.binsize > 2 else data)

    def many(self, data, subtract = True) -> np.ndarray:
        "call on all tables"
        itr = (self.__get(j) for j in data) if self.binsize > 2 else data
        fcn = getattr(np, self.mode.value)
        res = np.fromiter((-fcn(i) for i in itr), dtype = np.float32)
        if subtract:
            return np.subtract(res, np.nanmedian(res), out = res)
        else:
            return res

    def __call__(self, data) -> np.ndarray:
        if isinstance(data, np.ndarray) and np.isscalar(data[0]):
            return self.one(data)
        else:
            return self.many(data)

    @classmethod
    def run(cls, data, **kwa):
        "runs the algorithm"
        return cls(**kwa)(data)

class PhaseEdgeAlignment:
    """
    Functor which an array of biases computed as the extremum of provided ranges.
    Biases are furthermore centered at zero around their median

    Attributes:

    * *window*: the width on which to compute a median.
    * *edge*: the edge to use: left or right
    """
    class Mode(Enum):
        "Computation modes EdgeAlignment."
        left  = 'left'
        right = 'right'

    window     = 15
    edge       = Mode.left
    percentile = 75.
    @initdefaults(frozenset(locals()))
    def __init__(self, **_):
        pass

    def __call__(self, data, subtract = True) -> np.ndarray:
        sli = (slice(self.window) if self.edge is self.Mode.left else
               slice(-self.window, None))
        res = np.fromiter((-np.percentile(i[sli], self.percentile) for i in data),
                          dtype = 'f4')

        if subtract:
            return np.subtract(res, np.nanmedian(res), out = res)
        else:
            return res

    @classmethod
    def run(cls, data, **kwa):
        "runs the algorithm"
        return cls(**kwa)(data)

class CorrelationAlignment:
    """
    Finds biases which correlate best a cycle's histogram to the histogram of
    all cycles. This repeated multiple times with the latter histogram taking
    prior biases into account.

    Biases are furthermore centered at zero around their median

    Attributes:

    * *oversampling*: the amount by which histograms are oversampled. This is
        the computation's precision
    * *maxcorr*: max amount by which a cycle may be translated.
    * *nrepeats*: the number of times the procedure is repeated
    * *kernel*: the smoothing kernel to use on the data
    """
    oversampling = 5
    maxcorr      = 4
    nrepeats     = 6
    kernel       = KernelConvolution() # type: Optional[KernelConvolution]
    __DEFAULTS   = 'oversampling', 'maxcorr', 'nrepeats'
    @initdefaults(__DEFAULTS)
    def __init__(self, **kwa):
        self.kernel = kwa.get("kernel", KernelConvolution(**kwa))

    @property
    def exact_oversampling(self) -> float:
        "The exact oversampling used: int(oversampling)//2 * 2 +1"
        return (int(self.oversampling)//2) * 2 + 1

    @kwargsdefaults(__DEFAULTS+('kernel',))
    def __call__(self, data: Iterable[np.ndarray], **kwa) -> np.ndarray:
        if len(kwa):
            kwa = {i.replace('kernel_', ''):j for i, j in kwa.items()}
            with changefields(self.kernel, kwa):
                return self.__call__(data)

        if isinstance(data, Iterator):
            data = tuple(data)
        data    = cast(Sequence[np.ndarray], data)

        osamp   = (int(self.oversampling)//2) * 2 + 1
        maxcorr = self.maxcorr*osamp

        kern  = self.kernel(oversampling = osamp, range = 'same')
        ref   = np.empty((max(len(i) for i in data)*osamp+maxcorr*2,), dtype = 'f4')
        hists = []

        cur   = ref[:-2*maxcorr]
        for rng in data:
            cur.fill(0.)
            cur[osamp//2:len(rng)*osamp:osamp] = rng
            hists.append(kern(cur))

        bias = np.full((len(hists),), maxcorr+osamp//2, dtype = np.int32)
        for _ in range(self.nrepeats):
            ref.fill(0.)
            for start, hist in zip(bias, hists):
                ref[start:start+len(hist)] += hist

            bias  = np.fromiter((np.argmax(np.correlate(ref, hist)) for hist in hists),
                                dtype = np.float32, count = len(hists))
            bias += maxcorr+osamp//2-(bias.max()+bias.min())//2

        ref = bias/osamp
        return np.subtract(ref, np.median(ref), out = ref)

    @classmethod
    def run(cls, data: Iterable[np.ndarray], **kwa):
        "runs the algorithm"
        return cls()(data, **kwa)

Alignment = Union[ExtremumAlignment, CorrelationAlignment]

File: src/eventdetection/alignment.cpp
#include <vector>
#include <algorithm>
#include "signalfilter/accumulators.hpp"
#include "eventdetection/alignment.h"

namespace eventdetection { namespace alignment {
namespace {
    using namespace signalfilter::stats;

    info_t _ebincompute(size_t binsize, DataInfo const & data, bool ismin)
    {
        info_t out(data.ncycles);
        auto thr = (ismin ? 1.0f: -1.0f)* std::numeric_limits<float>::max();
        auto fcn = [ismin](float a, float b) { return ismin == (a < b) ? a : b; };
        for(size_t i = 0; i < data.ncycles; ++i)
        {
            auto j    = size_t(data.first[i]), e = size_t(data.last[i]);
            auto ptr  = data.data+j;
            auto minv = thr;
            for(; j+binsize < e; j += binsize, ptr += binsize)
                minv = fcn(minv, nanmedian(std::vector<float>(ptr, ptr+binsize)));
            if(j < e)
                minv = fcn(minv, nanmedian(std::vector<float>(ptr, data.data+e)));

            out[i] = -minv;
        }
        return out;
    }

    template <typename T>
    info_t _ecompute(DataInfo const & data, T && fcn)
    {
        info_t out(data.ncycles);
        auto   ptr(data.data);
        for(size_t i = 0; i < data.ncycles; ++i)
            out[i] = -fcn(ptr+data.first[i], ptr+data.last[i]);
        return out;
    }
}

info_t ExtremumAlignment::compute(DataInfo const && data) const
{
    info_t out;
    if(mode == ExtremumAlignment::median)
        out = _ecompute(data,
                        [](float const * a, float const *b)
                        { return signalfilter::stats::nanmedian(a, b); });
    else if(binsize >= 2)
        out = _ebincompute(binsize, data, mode == ExtremumAlignment::min);
    else if(mode == ExtremumAlignment::min)
        out = _ecompute(data,
                        [](float const * a, float const *b)
                        { return std::min_element(a, b)[0]; });
    else
        out = _ecompute(data,
                        [](float const * a, float const *b)
                        { return std::max_element(a, b)[0]; });
    return out;
}

info_t PhaseEdgeAlignment::compute(DataInfo const && data) const
{
    auto  ptr  (data.data);
    auto  first(data.first), last (data.last);
    float perc ((float) percentile);
    int   sz   ((int) window);

    info_t out(data.ncycles);
    if(mode == PhaseEdgeAlignment::left)
        for(size_t i = 0; i < data.ncycles; ++i)
        {
            std::vector<float> tmp(ptr+first[i], ptr+std::min(first[i]+sz, last[i]));
            out[i] = -signalfilter::stats::nanpercentile(tmp.data(),
                                                         tmp.data()+tmp.size(),
                                                         perc);
        }
    else
        for(size_t i = 0; i < data.ncycles; ++i)
        {
            std::vector<float> tmp(ptr+std::max(first[i], last[i]-sz), ptr+last[i]);
            out[i] = -signalfilter::stats::nanpercentile(tmp.data(),
                                                         tmp.data()+tmp.size(),
                                                         perc);
        }
    return out;
}

void translate(DataInfo const && data, bool del, float * ptr)
{
    if(data.ncycles == 0 || data.size == 0)
        return;

    using signalfilter::stats::nanmedian;
    auto delta(data.data);

    auto apply = [&](float tmp, size_t r1, size_t r2)
                  {
                      if(!std::isfinite(tmp) && !del)
                          tmp = -nanmedian(std::vector<float>(ptr+r1, ptr+r2));
                      if(std::isfinite(tmp))
                          for(size_t j = r1; j < r2; ++j)
                              ptr[j] += tmp;
                      else
                          std::fill(ptr+r1, ptr+r2, std::numeric_limits<float>::quiet_NaN());
                  };
    for(size_t i = 0u, e = data.ncycles-1; i < e; ++i)
        apply(delta[i], data.first[i], data.first[i+1]);
    apply(delta[data.ncycles-1], data.first[data.ncycles-1], data.size);
}

void medianthreshold(DataInfo const && data, float minv, float * bias)
{
    using signalfilter::stats::median;
    auto ptr(data.data);

    std::vector<float> values;
    std::vector<float> meds;
    std::vector<float> spread;
    std::vector<float> tmp;
    for(size_t i = 0u, e = data.ncycles; i < e; ++i)
    {
        tmp.resize(0);
        for(auto xval = ptr+data.first[i], eval = ptr+data.last[i]; xval != eval; ++xval)
            if(std::isfinite(*xval))
                tmp.push_back(*xval);

        if(tmp.size() < 2)
            values.push_back(std::numeric_limits<float>::quiet_NaN());
        else
        {
            auto med = median(tmp);
            for(auto & xval: tmp)
                xval = std::abs(xval-med);

            meds.push_back(med+bias[i]);
            values.push_back(med+bias[i]);
            spread.push_back(median(tmp));
        }
    }

    if(meds.size() < 2)
        return;

    auto med = median(meds) - median(spread) - minv;
    if(!std::isfinite(med))
        return;


    for(size_t i = 0u, e = data.ncycles; i < e; ++i)
        if(std::isfinite(values[i]) && values[i] < med)
            bias[i] = std::numeric_limits<float>::quiet_NaN();
}
}}
